magda-int-test/src/test/scala/au/csiro/data61/magda/api/DataSetSearchSpec.scala

```diff
@@ -110,6 +113,63 @@ class DataSetSearchSpec extends BaseSearchApiSpec with RegistryConverters {
       }
     }
 
+    describe("searching for a dataset should return that datasets contains the keyword & it's synonyms") {
+      it("for synonyms group 300032733 `agile`, `nimble`, `quick` & `spry`") {
+        case class GenResult(searchKeyword: String, synonym: String, datasetWithSynonym: DataSet)
+
+        val synonyms = List("agile", "nimble", "quick", "spry")
+        val pairs = synonyms.map(s => {
+          val listWithItself = synonyms.filterNot(_===s)
+          (s, listWithItself)
+        })
+
+        val cache = scala.collection.mutable.HashMap.empty[String, List[_]]
+        val randomDatasets = Gen.listOfN(20, Generators.dataSetGen(cache)).retryUntil(_ => true).sample.get
+
+        val synonymTestData = synonyms.map{
+          case(searchKeyword) =>
+            val gen = for {
+              synonym <- Gen.oneOf(synonyms.filter(_!==searchKeyword))
+              dataset <- Generators.dataSetGen(scala.collection.mutable.HashMap.empty)
+              datasetWithSynonym = dataset.copy( description = Some(synonym))
+            } yield GenResult(searchKeyword, synonym, datasetWithSynonym)
+            gen.retryUntil(_ => true).sample.get
+        }
+
+        val synonymDataset = synonymTestData.map(_.datasetWithSynonym)
+
+        val searchKeywordGen = for {
+          genResult <- Gen.oneOf(synonymTestData)
```

Comment:
Is it possible to put the `randomDatasets` and `synonymTestData` into this generator instead of using `.retryUntil(_ => true).sample.get` ? Or was that too slow?

Reply:
@AlexGilleran 
I am actually not very familiar with the generator API.
Those lines are copies of your `acronym` test cases example.
What I was trying to do is:
- Generate 20 random datasets as `background noise`
- Use built-in synonyms list to generate a few datasets contains synonyms words
- Put all datasets above into `elasticsearch` and search synonym keyword
Do you suggest putting both datasets generation into searchKeywordGen? Why do we want to do that?

Reply:
`.retryUntil(_ => true).sample.get` is a bit dodgy because it breaks ScalaCheck's ability to shrink input... say you generate `randomDatasets` as `List(a, b, c)`. If you're using the generator inside a `forAll`, then if the test fails Scalacheck is able to try it with `randomDatasets` = `List(a, b)`, `List(b, c)` etc until it finds the smallest possible failing case. 

I guess this is ok though, because if you had all the generated datasets in the `forAll` you'd have to insert a new index 100 times instead of once, which'd be too slow 🤷‍♂️

Reply:
Thanks a lot for your comments 👍 
Good to know the difference & I'd better also have a read on ScalaCheck documents